table/format.cc

```diff
@@ -206,74 +210,81 @@ Footer& Footer::set_table_magic_number(uint64_t magic_number) {
-  assert(IsLegacyFooterFormat(magic) == (fv == 0));
+constexpr size_t kFooterPart2Size = 2 * BlockHandle::kMaxEncodedLength;
+}  // namespace
 
-  ChecksumType ct = checksum_type();
+void FooterBuilder::Build(uint64_t magic_number, uint32_t format_version,
+                          uint64_t footer_offset, ChecksumType checksum_type,
+                          const BlockHandle& metaindex_handle,
+                          const BlockHandle& index_handle) {
+  (void)footer_offset;  // Future use
+
+  assert(magic_number != Footer::kNullTableMagicNumber);
+  assert(IsSupportedFormatVersion(format_version));
 
-  // Allocate destination data and generate parts 1 and 3
-  const size_t original_size = dst->size();
   char* part2;
-  if (fv > 0) {
-    dst->resize(original_size + kNewVersionsEncodedLength);
-    char* part1 = &(*dst)[original_size];
-    part2 = part1 + 1;
-    char* part3 = part2 + 2 * BlockHandle::kMaxEncodedLength;
-    assert(&(*dst)[dst->size() - 1] + 1 - part3 == /* part 3 size */ 12);
+  if (format_version > 0) {
+    slice_ = Slice(data_.data(), Footer::kNewVersionsEncodedLength);
     // Generate parts 1 and 3
-    part1[0] = ct;
-    EncodeFixed32(part3, fv);
-    EncodeFixed64(part3 + 4, magic);
+    char* cur = data_.data();
+    // Part 1
+    *(cur++) = checksum_type;
+    part2 = cur;
+    // Skip over part 2 for now
+    cur += kFooterPart2Size;
+    // Part 3
+    EncodeFixed32(cur, format_version);
+    cur += 4;
+    EncodeFixed64(cur, magic_number);
+    assert(cur + 8 == slice_.data() + slice_.size());
   } else {
-    dst->resize(original_size + kVersion0EncodedLength);
-    part2 = &(*dst)[original_size];
-    char* part3 = part2 + 2 * BlockHandle::kMaxEncodedLength;
-    assert(&(*dst)[dst->size() - 1] + 1 - part3 == /* part 3 size */ 8);
+    slice_ = Slice(data_.data(), Footer::kVersion0EncodedLength);
     // Legacy SST files use kCRC32c checksum but it's not stored in footer.
-    assert(ct == kNoChecksum || ct == kCRC32c);
-    // Generate part 3 (part 1 empty)
-    EncodeFixed64(part3, magic);
+    assert(checksum_type == kNoChecksum || checksum_type == kCRC32c);
+    // Generate part 3 (part 1 empty, skip part 2 for now)
+    part2 = data_.data();
+    char* cur = part2 + kFooterPart2Size;
+    // Use legacy magic numbers to indicate format_version=0, for
+    // compatibility. No other cases should use format_version=0.
+    EncodeFixed64(cur, DownconvertToLegacyFooterFormat(magic_number));
+    assert(cur + 8 == slice_.data() + slice_.size());
   }
 
-  // Generate Part2
-  // Variable size encode handles (sigh)
-  part2 = metaindex_handle_.EncodeTo(part2);
-  /*part2 = */ index_handle_.EncodeTo(part2);
-
-  // remainder of part2 is already zero padded
+  {
+    char* cur = part2;
+    char* end = part2 + kFooterPart2Size;
+    cur = metaindex_handle.EncodeTo(cur);
+    cur = index_handle.EncodeTo(cur);
+    // Zero pad remainder
+    std::fill(cur, end, char{0});
```

Comment:
Nit: Should there be an assert here to make sure part2 does not overflow into part3?

Reply:
We compute the start of part3 a few lines up using `part2 + kFooterPart2Size`. It perhaps makes sense to keep the `part3` variable. It was lost in sweeping out unnecessary variables.